### ISSUE TITLE
$route['404_override'] controller/method is called in Bootstrap

### ISSUE THREAD

**Issue by PhoenixFnX** [kenjis/ci-phpunit-test#58]
When I follow the head tutorial, I can't get the bootstrap to work when there is a redirect() in my default controller. My Auth/index was redirecting to Auth/login.
I removed it and called directly $this->login() in Auth/index().

**Comment by kenjis** [comment]
I can't reproduce it. Please show me how to reproduce.

**Comment by PhoenixFnX** [comment]
Ok let me try to tell you how to reproduce:
First, here is the stacktrace when the bootstrap file is ran:
    Bootstrap.php:331
    CIPHPUnitTest:65
    CodeIgniter.php:514 : Here is called the default controller/Method

In my routes.php I have the following : 
$route['default_controller'] = 'auth';

Therefore I get the function Auth/index executed in the bootstrap, which was containing a redirect() to Auth/login. 

The assertRedirect is useful for tests but I'm still in bootstrap there and there is not yet a hook on redirect method.

**Comment by kenjis** [comment]
Do you have catch-all route?

**Comment by PhoenixFnX** [comment]
Nope, I don't know that feature.
My whole routes.php file is 
<pre>
$route['default_controller'] = 'auth';
$route['404_override'] = '/';
$route['translate_uri_dashes'] = FALSE;
</pre>

**Comment by kenjis** [comment]
> Bootstrap.php:331
CIPHPUnitTest:65
CodeIgniter.php:514 : Here is called the default controller/Method

Normally `CodeIgniter.php:514` does not run. Because ci-phpunit-test call `_dummy/_dummy` (changed `$_SERVER['argv']`) in bootstrap and it is 404. 

So Line 476 `show_404($RTR->directory.$class.'/'.$method);` is called.

Why do you go to line 514?

**Comment by kenjis** [comment]
Ah! Because of `404_override`?

**Comment by PhoenixFnX** [comment]
Hum maybe... 
	$RTR =& load_class('Router', 'core', isset($routing) ? $routing : NULL);
	$class = ucfirst($RTR->class);
	$method = $RTR->method;

At this 514 line:
	$class=="Auth"
	$method=="index"

**Comment by kenjis** [comment]
By the way, you should set like `$route['404_override'] = 'controller/method';`.

Your `$route['404_override'] = '/';` is apparently wrong.

**Comment by PhoenixFnX** [comment]
I just want the show_404 method to be called but anyway, I created a controller/method to see if this is the problem, I is not : 
![image](https://cloud.githubusercontent.com/assets/993122/10163115/d22d094e-66b0-11e5-9aef-6bd3ac54974a.png)

I don't see any override with _dummy/_dummy in bootstrap, where is it ?

**Comment by kenjis** [comment]
See https://github.com/kenjis/ci-phpunit-test/blob/master/application/tests/_ci_phpunit_test/CIPHPUnitTest.php#L19-L22

I don't understand why your values are Auth and index.

**Comment by PhoenixFnX** [comment]
Is it normal that argv is still the same here:
![image](https://cloud.githubusercontent.com/assets/993122/10163303/340c778e-66b2-11e5-80f8-a870a3dde03d.png)

**Comment by PhoenixFnX** [comment]
Here in the CIPHPUnitTest.php:72, there is a $_SERVER restoration:
<pre>
// Restore $_SERVER
		$_SERVER = $_server_backup;
</pre>

**Comment by kenjis** [comment]
> Where is the $_SERVER["argv"] is it supposed to be used ?

CodeIgniter uses it to get route. See `URI::_parse_argv()`.

**Comment by kenjis** [comment]
Try this branch: https://github.com/kenjis/ci-phpunit-test/tree/fix-404_override

**Comment by PhoenixFnX** [comment]
with this return, yes there is no problem :D

**Comment by PhoenixFnX** [comment]
@kenjis I had a patch in the URI file for CLI uses, I was removing argv parameters :/
I removed the patch and now the argv is correctly read !
I removed your patch to get the try catch back and the show_404 exception is now correctly caught.

**Comment by PhoenixFnX** [comment]
It's now failing in the new CI_Controller();
When initializing the Form_validation helper, the constructor tries to load the form helper, it fails because it doesn't find the load method in get_instance :/
<pre>
	public function __construct($rules = array()) 
	{
		$this->CI =& get_instance();
[...]
		// Automatically load the form helper
		$this->CI->load->helper('form');

		log_message('info', 'Form Validation Class Initialized');
	}
</pre>

**Comment by PhoenixFnX** [comment]
This is probably because I have a MY_Form_validation class

**Comment by kenjis** [comment]
What's `CI_Controller::$loadFilename`?

**Comment by PhoenixFnX** [comment]
Actually, you try to instanciate a second time the Controller class calling new CI_Controller();

It was loaded the first time at line 500 in CodeIgniter.php. Autoload libraries were loaded too (including form_validation which loads a helper using get_instance()->load object which was correctly initialised).

The problem is with this first loop in Controller.php which uses the is_loaded() static method. Because it is static, it remembers that Form_validation was loaded one time. Therefore it tries to instantiate it again with load_class("Form_validation") and call the constructor, and call $this->CI->load which is not yet initialized !

**Comment by PhoenixFnX** [comment]
I think you should reset the is_loaded() before calling new CI_Controller(); in CIPHPUnitTest.php:68

**Comment by PhoenixFnX** [comment]
To reproduce, I think you just should add form_validation in the autoload libraries

**Comment by PhoenixFnX** [comment]
only resetting the is_loaded() array is not working but the reset_instance() function does the job :+1: 

<pre>
try {
			// Request to 404 route
			// This is needed for not to call Welcome::index()
			// If controller Welcome is called in bootstrap, we can't test
			// the same name sub controller Welcome even when we use
			// `@runInSeparateProcess` and `@preserveGlobalState disabled`
			require_once BASEPATH . 'core/CodeIgniter.php';
		} catch (CIPHPUnitTestShow404Exception $e) {
			// Catch 404 exception
			reset_instance();
			new CI_Controller();
		}
</pre>

**Comment by kenjis** [comment]
After all, if we have `404_override`, the controller is called in Bootstrap. And you see the output.
It should not be called.

I've fixed it on `master` branch.

**Comment by PhoenixFnX** [comment]
I installed your framework through composer, how do I upgrade it with this last commit ?
composer update kenjis/ci-phpunit-test = Nothing to install or update

**Comment by kenjis** [comment]
`master` branch is not released.

If you want to use it:

 1) Change the version in composer.json.

~~~diff
--- a/composer.json
+++ b/composer.json
@@ -6,6 +6,6 @@
        },
        "require-dev": {
                "mikey179/vfsStream": "1.1.*",
-               "kenjis/ci-phpunit-test": "^0.8.0"
+               "kenjis/ci-phpunit-test": "1.0.x@dev"
        }
 }
~~~

 2) Run composer update.  

~~~
$ composer update kenjis/ci-phpunit-test
~~~

 3) Install to `application/tests` folder.

~~~
$ php vendor/kenjis/ci-phpunit-test/update.php
~~~

**Comment by kenjis** [comment]
I've just released v0.8.1.

**Comment by kenjis** [comment]
I believe this bug was fixed.
